Add fn key combinations to emit international keyboard keys

Jules-Bertholet opened the pull request:
Maps fn+Z to the ISO 102nd key, fn+/ to the extra key left of right shift on Brazilian and Japanese layouts, and fn+= to the JIS Yen key. This allows full use of ISO and Brazilian layouts on ANSI, ISO, or JIS keyboard hardware. (Alternately, users can remap these keys in the OS to whatever function they prefer.)

Jules-Bertholet committed:
fwk: add fn key combinations for row B end keys

Maps fn+Z to the ISO 102nd key, and fn+/ to the extra key
left of right shift on Brazilian and Japanese layouts.
This allows full use of ISO and Brazilian layouts on
ANSI, ISO, or JIS keyboard hardware. (Alternately,
users can remap these keys in the OS to whatever function
they prefer.)

Signed-off-by: Jules Bertholet <julesbertholet@quoi.xyz>

Jules-Bertholet committed:
fwk: map fn+= to Yen key

Maps fn+= to the JIS Yen key to the left of Backspace
(also the Won key on certain Korean layouts).
This allows fuller use of these layouts on non-JIS hardware;
alternately, users can remap this keys in the OS to whatever
function they prefer.

Signed-off-by: Jules Bertholet <julesbertholet@quoi.xyz>

JohnAZoidberg commented:
Is there precedent for other keyboards to do this?
Do you use this and find it useful?

Jules-Bertholet commented:
I don’t know of any existing keyboard that does this, no. But yes, I would use it if it was available (likely by remapping the keys to custom actions/layers).

JohnAZoidberg commented:
Hmm, if this mapping is not standard, I don't think it would be very popular to use. It's more like a personal keyboard layout that you made for yourself.

Best would be if we can get opinions from Brazilians or Japanese people. We're not shipping to these countries yet, we'll keep it in mind for when we launch there.

Jules-Bertholet commented:
The fn+Z to the ISO 102nd key mapping is relevant to anyone with ANSI (US QWERTY) keyboard hardware that occasionally uses an ISO layout (all European layouts, Chinese, Korean, etc).

Also, these mappings are not chosen randomly or by my personal preference, but based on what is the closest key to the one being emulated. For example, Z is directly to the right of Left Shift on non-ISO layouts, so fn+Z maps to the 102nd key that is directly to the right of Left Shift on ISO layouts. The slash key is to the left of Right Shift on non-JIS or Brazilian layouts, so fn+/ maps to the extra key to the left of Right Shift on JIS and Brazilian layouts. The = key is to the left of Backspace on non-JIS, so fn+= maps to the JIS Yen key that is to the left of Backspace. See the blue arrows on the image below:

![iso keyboard](https://github.com/user-attachments/assets/8b980218-096e-4660-97b3-0ab2ffe92ec6)

Jules-Bertholet committed:
fwk: add fn key mappings for JIS IME keys

fn+C is mapped to Muhenkan,
fn+M is mapped to Henkan,
and fn+, is mapped to Katakana/Hiragana.

Signed-off-by: Jules Bertholet <julesbertholet@quoi.xyz>

Jules-Bertholet commented:
I’ve added mappings to the three JIS IME keys. Here is the full JIS set (only Z → ISO 102nd is omitted):

![fwk jis](https://github.com/user-attachments/assets/7552c7a5-b554-4d3c-bc4a-6af17ff10803)

JohnAZoidberg commented:
> Also, these mappings are not chosen randomly or by my personal preference, but based on what is the closest key to the one being emulated

Sure, I understand and that makes sense. But if a keyboard layout is not standard, people are not used to it and need to train their muscle memory for it - even if it's slightly different.
I do think this is a great patch for those who want it but I'm not convinced yet if it's useful to merge for everyone.

I just talked to a japanese developer and he said that they expect the Yen key on japanese keyboards to result in a backslash (\) scancode. He said sending Yen keycode is a common mistake by hardware manufacturers. Operating systems will map backslash to Yen symbol based on necessity and OS configured layout.

Jules-Bertholet commented:
>  I do think this is a great patch for those who want it but I'm not convinced yet if it's useful to merge for everyone.

I think I now agree, #55 is a better approach compared to this.

> I just talked to a japanese developer and he said that they expect the Yen key on japanese keyboards to result in a backslash () scancode.

Framework’s own JIS keyboard does not do this, the Yen key sends the Yen scancode. And for good reason: the key to the immediate left of `Enter` on the home row already sends the scancode that, on ANSI-QWERTY, maps to backslash.
